pkg/cmd/cli/cmd/edit.go

```diff
@@ -0,0 +1,333 @@
+		// TODO: simplify by adding a ResourceBuilder mode
+		defaultVersion := clientConfig.Version
+		// the outermost object will be converted to the output-version
+		version := cmdutil.OutputVersion(cmd, defaultVersion)
+		versions := []string{version, latest.Version}
+		infos, _ := r.Infos()
+		for _, info := range infos {
+			versions = append(versions, info.Mapping.APIVersion)
+		}
+
+		// TODO: add a new ResourceBuilder mode for Object() that attempts to ensure the objects
+		//   are in the appropriate version if one exists (and if not, use the best effort).
+		// TODO: ensure api-version is set with the default preferred api version by the client
+		//   builder on initialization
+		// TODO: add an annotating YAML printer that can print inline comments on each field,
+		//   including descriptions or validation errors
+		printer := kubectl.NewVersionedPrinter(p, kapi.Scheme, versions...)
+
+		// generate the file to edit
+		buf := &bytes.Buffer{}
+		if err := results.header.WriteTo(buf); err != nil {
+			return err
+		}
+		if err := printer.PrintObj(obj, buf); err != nil {
+			return err
+		}
+		original := buf.Bytes()
+
+		// launch the editor
+		edit := editor.NewDefaultEditor()
+		edited, file, err := edit.LaunchTempFile("", "osc-edit-", buf)
+		if err != nil {
+			return err
+		}
+		glog.V(4).Infof("User edited:\n%s", string(edited))
+		lines, err := hasLines(bytes.NewBuffer(edited))
+		if err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+		if bytes.Equal(original, edited) {
+			os.Remove(file)
+			fmt.Fprintln(cmd.Out(), "Edit cancelled, no changes made.")
+			return nil
+		}
+		if !lines {
+			os.Remove(file)
+			fmt.Fprintln(cmd.Out(), "Edit cancelled, saved file was empty.")
+			return nil
+		}
+
+		results = editResults{}
+
+		// parse the edited file
+		info, err := rmap.InfoForData(edited, "edited-file")
+		if err != nil {
+			results.header.reasons = append(results.header.reasons, editReason{
+				head: fmt.Sprintf("The edited file had a syntax error: %v", err),
+			})
+			continue
+		}
+
+		err = resource.NewFlattenListVisitor(info, rmap).Visit(func(info *resource.Info) error {
+			data, err := info.Mapping.Codec.Encode(info.Object)
+			if err != nil {
+				return err
+			}
+			updated, err := resource.NewHelper(info.Client, info.Mapping).Update(info.Namespace, info.Name, false, data)
+			if err != nil {
+				fmt.Fprintln(cmd.Out(), results.AddError(err, info))
+				return nil
+			}
+			info.Refresh(updated, true)
+			fmt.Fprintf(out, "%s/%s\n", info.Mapping.Resource, info.Name)
+			return nil
+		})
+		if err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+
+		if results.retryable > 0 {
+			fmt.Fprintf(cmd.Out(), "You can run `%s update -f %s` to try this update again.\n", fullName, file)
+			return errExit
+		}
+		if results.conflict > 0 {
+			fmt.Fprintf(cmd.Out(), "You must update your resource version and run `%s update -f %s` to overwrite the remote changes.\n", fullName, file)
```

Comment:
I think fixing conflict will just be a follow up - 3 way merge with the work Sam did should allow us to do that.

----- Original Message -----
> > +				return nil
> > +			}
> > +			info.Refresh(updated, true)
> > +			fmt.Fprintf(out, "%s/%s\n", info.Mapping.Resource, info.Name)
> > +			return nil
> > +		})
> > +		if err != nil {
> > +			return preservedFile(err, file, cmd.Out())
> > +		}
> > +
> > +		if results.retryable > 0 {
> > +			fmt.Fprintf(cmd.Out(), "You can run `%s update -f %s` to try this
> > update again.\n", fullName, file)
> > +			return errExit
> > +		}
> > +		if results.conflict > 0 {
> > +			fmt.Fprintf(cmd.Out(), "You must update your resource version and run
> > `%s update -f %s` to overwrite the remote changes.\n", fullName, file)
> 
> Also, in a multi-edit scenario, this advice will fail again because one edit
> took place.  Is there a reasonable way to strip the file down to only those
> elements that failed?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1758/files#r28531213